# Recursive Function

In [1]:
# def factorial(num):
#     if num==1:
#             return 1
#     else:
#         return num * factorial(num - 1)


# factorial(5)

120

Logic:
    factorial(5)----> 5 * fact(4)   (Here 4 is comapred to if statement)
    
                              4 * fact(3)  (Here 3 is comapred to if statement)
                              
                                     3 * fact(2)   (Here 2 is comapred to if statement)
                                     
                                            2 * fact(1)   (Here 1 is comapred to if statement)
                                            
                                                   1           (Here 1 is comapred to if statement, **got a match**)
                    
        Atlast return would be 1 and then it will return a answer to the immediate call function which is fact(1)
        So, 2 * fact(1)= 2 * 1= 2. Then, 2 will return ans to its immediate call which is 
            3 * fact(2)= 3 * 2= 6 and so on,
            4 * fact(3)= 4 * 6= 24,
            5 * fact(4)= 5 * 24= 120.Got the Answer, because no call function present.

In [4]:
#logic
# mid= 9/2= 4 = 500 == 700 

# [600,700,800,900]

# mid= 4/2 = 2 = 800 == 700

# [600,700]

# mid = 2/2 = 1 =700 == 700 return True

def binary_search(l,key):
    if len(l)==0:
        return False
    else:
        mid= len(l)//2
        if l[mid]==key:
            return True
        elif key<l[mid]:
            return binary_search(l[:mid],key)
        else:
            return binary_search(l[mid+1:],key)
print(__name__)  # defaultly python is keeping the name of module as main 

if __name__ == '__main__' :
    l = [100,200,300,400,500,600,700,800,900]
    key = 700
    binary_search(l,500)



__main__
